pkg/parameters/aws/aws.go

```diff
@@ -0,0 +1,213 @@
+	config := aws.NewConfig().WithHTTPClient(httpClient).WithRegion(region)
+	return config
+}
+
+func NewAWSParameters(base string, config *aws.Config) (*AWSParameters, error) {
+	if config == nil {
+		config = getAWSConfig()
+	}
+
+	instanceID := detectInstanceID()
+	if instanceID == "" {
+		return nil, fmt.Errorf("failed to detect AWS instance ID, not running on AWS?")
+	}
+
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	ec2Client := ec2.New(sess, config)
+
+	sess, err = session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	ssmClient := ssm.New(sess, config)
+
+	return &AWSParameters{
+		base: filepath.Join(base, instanceID),
+		ec2:  ec2Client,
+		ssm:  ssmClient,
+	}, nil
+}
+
+func unmarshalParameters(params map[string]string) (map[string]string, error) {
+	// This is either one single SSM parameter if it fit (<=4096 bytes), or
+	// chunks of the serialized config e.g. "config-0", "config-1", etc. We
+	// need to assemble the chunks in order, de-serialize and re-create the
+	// original config map.
+	if len(params) == 0 {
+		return nil, fmt.Errorf("got no SSM parameters")
+	}
+
+	// The whole map in one parameter.
+	if len(params) == 1 {
+		for k, v := range params {
+			configMap := make(map[string]string)
+			klog.V(2).Infof("unmarshaling SSM parameter %q", k)
+			err := yaml.Unmarshal([]byte(v), &configMap)
+			if err != nil {
+				return nil, err
+			}
+			return configMap, nil
+		}
+	}
+
+	// Multiple chunks.
+	paramList := make([]string, len(params))
+	for k, v := range params {
+		parts := strings.SplitN(k, "-", 2)
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("invalid SSM parameter chunk key: %s", k)
+		}
+		n, err := strconv.ParseUint(parts[1], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		if int(n) >= len(paramList) {
+			return nil, fmt.Errorf("invalid SSM parameter chunk key: %s", k)
+		}
+		paramList[n] = v
+	}
+	yml := ""
+	for i := range paramList {
+		yml = yml + paramList[i]
+	}
+	configMap := make(map[string]string)
+	klog.V(2).Infof("unmarshaling %d SSM parameter chunks, %d bytes", len(paramList), len(yml))
+	err := yaml.Unmarshal([]byte(yml), &configMap)
+	if err != nil {
+		return nil, err
+	}
+	return configMap, nil
+}
+
+func (a *AWSParameters) getParameter(name string) string {
+	path := filepath.Join(a.base, name)
+	in := &ssm.GetParameterInput{
+		Name:           aws.String(path),
+		WithDecryption: aws.Bool(true),
+	}
+	out, err := a.ssm.GetParameter(in)
+	if err != nil {
+		klog.V(2).Infof("getting SSM parameter %s: %v", name, err)
+		return ""
+	}
+	if out.Parameter == nil {
+		klog.V(2).Infof("got nil SSM parameter value for %s", name)
+		return ""
+	}
+	return aws.StringValue(out.Parameter.Value)
+}
+
+func (a *AWSParameters) GetAllParameters() (map[string]string, error) {
+	params := make(map[string]string)
+
+	name := ssmParameterBaseName
+	value := a.getParameter(name)
+	if value != "" {
+		params[name] = value
+	}
+
+	for i := 0; i < ssmMaxResults; i++ {
```

Comment:
Just to clarify: we cannot use `ssm.GetParameters` here, right? Is it because parameters can be bigger than max response key length?

Reply:
That's what I used first, but the way IAM policies are set up, it gets a deny from IAM even for non-existent parameters. E.g.:

    aws --region us-west-2 ssm get-parameters --names /kip/cells/i-06792ed23f773a037/config
    {
        "InvalidParameters": [],
        "Parameters": [
            {
                "ARN": "arn:aws:ssm:us-west-2:689494258501:parameter/kip/cells/i-06792ed23f773a037/config",
                "Value": "<redacted>",
                "DataType": "text",
                "Name": "/kip/cells/i-06792ed23f773a037/config",
                "Version": 1,
                "LastModifiedDate": 1607474007.126,
                "Type": "SecureString"
            }
        ]
    }
    
    aws --region us-west-2 ssm get-parameters --names /kip/cells/i-06792ed23f773a037/config,/kip/cells/i-06792ed23f773a037/config-0,/kip/cells/i-06792ed23f773a037/config-1
    
    An error occurred (AccessDeniedException) when calling the GetParameters operation: User: arn:aws:sts::689494258501:assumed-role/devx-test_kip_cell_role/i-06792ed23f773a037 is not authorized to perform: ssm:GetParameters on resource: arn:aws:ssm:us-west-2:689494258501:parameter/kip/cells/i-06792ed23f773a037/config,/kip/cells/i-06792ed23f773a037/config-0,/kip/cells/i-06792ed23f773a037/config-1 with an explicit deny

So we have to do it one by one (or rethink the suggested IAM policy, but that's a deep rabbit hole). In the basic case, though, this is only 2 calls (for `config` and `config-0`, once `config-0` fails, we know that there are no more parameters).

Reply:
makes sense :+1: